xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,512 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE decoder
+//
+// The RLE decoder decompresses incoming stream of
+// (`symbol`, `count`) pairs, representing that `symbol` was
+// repeated `count` times in original stream. Output stream
+// should be equal to the RLE encoder input stream.
+// Both input and output channels use additional `last` flag
+// that indicates whether the packet ends the transmission.
+// Decoder in its current form only propagates last signal.
+// The behavior of the decoder is presented on the waveform below:
+//                      ──────╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥────
+// next evaluation      XXXXXX║ 0   ║ 1   ║ 2   ║ 3   ║ 4   ║ 5   ║ 6   ║ ...
+//                      ──────╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨────
+// do_recv                    ┌─────┐     ┌─────────────────┐           ┌────
+//                      ──────┘     └─────┘                 └───────────┘
+//                      ──────╥─────╥─────╥─────╥─────╥─────╥────────────────
+// symbol, count        XXXXXX║ A,2 ║XXXXX║ B,1 ║ B,1 ║ C,3 ║XXXXXXXXXXXXXXXX
+// (input channel)      ──────╨─────╨─────╨─────╨─────╨─────╨────────────────
+// last                                   ┌─────┐     ┌─────┐
+// (input channel)      ──────────────────┘     └─────┘     └────────────────
+//                      ╥─────╥───────────╥───────────╥─────────────────╥────
+// state.symbol         ║ 0   ║ A         ║ B         ║ C               ║ 0
+// (set state value)    ╨─────╨───────────╨───────────╨─────────────────╨────
+//                      ╥─────╥─────╥─────╥───────────╥─────╥─────╥─────╥────
+// state.count          ║ 0   ║ 1   ║ 0   ║ 0         ║ 2   ║ 1   ║ 0   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_common as rle_common
+
+type DecInData  = rle_common::EncData;
+type DecOutData = rle_common::PlainData;
+
+// structure to preserve the state of an RLE decoder
+struct RunLengthDecoderState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  // symbol to be repeated on output
+  symbol: bits[SYMBOL_WIDTH],
+  // count of symbols that has to be send
+  count: bits[COUNT_WIDTH],
+  // send last when repeat ends
+  last: bool,
+}
+ // RLE decoder implementation
+pub proc RLEDec<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
```

Comment:
rename to match the state naming convention?

Reply:
Names updated